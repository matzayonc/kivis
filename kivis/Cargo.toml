[package]
name = "kivis"
version = "0.1.0"
edition = "2024"

[dependencies]
thiserror = { version = "1.0.31" }
kivis-derive = { workspace = true }
serde = { version = "1.0.219", features = ["derive"] }
bcs = { version = "0.1.6" }
