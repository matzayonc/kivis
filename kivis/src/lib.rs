//! # Kivis Database
//!
//! A lightweight, type-safe database schema library for Rust with support for custom storage backends,
//! automatic indexing, and foreign key relationships.
//!
//! ## Features
//!
//! - **Type-safe schema generation**: Automatically derive database schemas from Rust structs
//! - **Generic storage backend support**: Work with any ordered key-value store
//! - **Automatic key generation and indexing**: Support for auto-increment keys and secondary indexes
//! - **Foreign key relationships**: Type-safe references between records
//! - **Layered cache architectures**: Compose multiple storage implementations
//!
//! ## Quick Start
//!
//! ```rust
//! use kivis::{Database, MemoryStorage, Record};
//!
//! #[derive(Record, serde::Serialize, serde::Deserialize, Debug)]
//! #[table(1)]
//! struct User {
//!     name: String,
//!     email: String,
//! }
//!
//! # fn main() -> Result<(), kivis::DatabaseError<kivis::MemoryStorageError>> {
//! let mut db = Database::new(MemoryStorage::new());
//! let user = User {
//!     name: "Alice".to_string(),
//!     email: "alice@example.com".to_string(),
//! };
//! let user_key = db.put(user)?;
//! # Ok(())
//! # }
//! ```

mod btreemap;
mod database;
mod errors;
mod manifest;
mod traits;
mod wrap;

pub use btreemap::{MemoryStorage, MemoryStorageError};
pub use database::Database;
pub use kivis_derive::Record;
pub use traits::*;

pub use crate::errors::{DatabaseError, InternalDatabaseError};
