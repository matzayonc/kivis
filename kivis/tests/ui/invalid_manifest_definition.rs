use kivis::{manifest, Scope};

struct User;
struct Post;

// This should fail - missing manifest name and colon
manifest![User, Post];

fn main() {}
