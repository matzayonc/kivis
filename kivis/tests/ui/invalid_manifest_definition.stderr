error: manifest! macro requires a manifest name followed by a colon. Use: manifest![ManifestName: Type1, Type2, ...]
 --> tests/ui/invalid_manifest_definition.rs:7:1
  |
7 | manifest![User, Post];
  | ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `manifest` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `Scope`
 --> tests/ui/invalid_manifest_definition.rs:1:23
  |
1 | use kivis::{manifest, Scope};
  |                       ^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
